consensus/consensus.go

```diff
@@ -1,65 +1,180 @@
+
+	// Update the consensus status
+	fc.status = consensusStatusInitialized
+
+	return nil
+}
+
+func (fc *FastHotStuffConsensus) UpdateChainTip( /* TODO */ ) {
 	// TODO
 }
 
-func (fc *FastHotStuffConsensus) HandleAcceptedBlock( /* TODO */ ) {
+func (fc *FastHotStuffConsensus) UpdateView( /* TODO */ ) {
 	// TODO
 }
 
-func (fc *FastHotStuffConsensus) HandleVoteMessage( /* TODO */ ) {
+func (fc *FastHotStuffConsensus) ProcessVoteMsg( /* TODO */ ) {
 	// TODO
 }
 
-func (pc *FastHotStuffConsensus) HandleTimeoutMessage( /* TODO */ ) {
+func (pc *FastHotStuffConsensus) ProcessTimeoutMsg( /* TODO */ ) {
 	// TODO
 }
 
-func (fc *FastHotStuffConsensus) HandleBlockProposal( /* TODO */ ) {
+func (fc *FastHotStuffConsensus) ConstructVoteQC( /* TODO */ ) {
 	// TODO
 }
 
+func (fc *FastHotStuffConsensus) ConstructTimeoutQC( /* TODO */ ) {
+	// TODO
+}
+
+// Sets the initial times for the the block construction and timeout timers and starts
+// the event loop building off of the current chain tip.
 func (fc *FastHotStuffConsensus) Start() {
+	fc.lock.Lock()
+	if fc.status != consensusStatusInitialized {
+		// Nothing to do here. The consensus instance is either already running or uninitialized.
+		fc.lock.Unlock()
+		return
+	}
+
+	// Update the consensus status to mark it as running.
+	fc.status = consensusStatusRunning
+
+	// Set the initial times for the the block construction and timeout timers
+	fc.nextBlockConstructionTimeStamp = time.Now().Add(fc.blockConstructionCadence)
+	fc.nextTimeoutTimeStamp = time.Now().Add(fc.timeoutBaseDuration)
+
+	// We can release the lock now that all state changes has been set up to start
+	// the event loop.
+	fc.lock.Unlock()
+
+	// Start the event loop
 	for {
 		select {
-		case <-time.After(time.Until(fc.nextBlockProposalTime)):
+		case <-time.After(time.Until(fc.nextBlockConstructionTimeStamp)):
 			{
 				// TODO
 			}
-		case <-time.After(time.Until(fc.nextTimeoutTime)):
+		case <-time.After(time.Until(fc.nextTimeoutTimeStamp)):
 			{
 				// TODO
 			}
-		case <-fc.internalTimersUpdated:
+		case <-fc.internalTimersUpdatedSignal:
 			{
 				// TODO
 			}
-		case <-fc.quit:
+		case <-fc.stopSignal:
 			{
-				// TODO
-				close(fc.quit)
+				fc.onStopSignal()
 				return
 			}
 		}
 	}
 }
 
+func (fc *FastHotStuffConsensus) IsInitialized() bool {
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	return fc.status != consensusStatusNotInitialized
+}
+
+func (fc *FastHotStuffConsensus) IsRunning() bool {
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	return fc.status == consensusStatusRunning
+}
+
 func (fc *FastHotStuffConsensus) Stop() {
-	fc.quit <- struct{}{}
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Grabbing the lock first and checking the status ensures that we can never push a stop
+	// signal once the channel has been closed. It's OK if we push multiple stop signals while
+	// the channel is still open.
+	if fc.status == consensusStatusRunning {
+		fc.stopSignal <- struct{}{}
```

Comment:
Could you add an exitGroup here? This will ensure that the Start's loop really finishes before we return from this function.